development/extensions/tutorial_authentication.rst

```diff
@@ -0,0 +1,287 @@
+service in addition to all normal requirements of an extension.
+
+Following is an example using a mock plugin, authext, by author imkingdavid.
+
+The class file
+++++++++++++++
+The provider class must implement the ``\phpbb\auth\provider\provider_interface`` in order to
+insure proper functionality. However, it is recommended to extend
+``\phpbb\auth\provider\base`` so as to not implement unneeded methods and to ensure
+that the provider will not break due to an update to the interface. An example
+authentication provider class is show below.
+
+.. code-block:: php
+
+    <?php
+
+    namespace imkingdavid\authext\auth\provider;
+
+    /**
+     * Database authentication provider for phpBB3
+     *
+     * This is for authentication via the integrated user table
+     */
+    class db2 extends \phpbb\auth\provider\base
+    {
+        /** @var \phpbb\db\driver\driver_interface $db */
+        protected $db;
+
+        /**
+         * Database Authentication Constructor
+         *
+         * @param \phpbb\db\driver\driver_interface $db
+         */
+        public function __construct(\phpbb\db\driver\driver_interface $db)
+        {
+            $this->db = $db;
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function login($username, $password)
+        {
+            // Auth plugins get the password untrimmed.
+            // For compatibility we trim() here.
+            $password = trim($password);
+
+            // do not allow empty password
+            if (!$password)
+            {
+                return array(
+                    'status'    => LOGIN_ERROR_PASSWORD,
+                    'error_msg' => 'NO_PASSWORD_SUPPLIED',
+                    'user_row'  => array('user_id' => ANONYMOUS),
+                );
+            }
+
+            if (!$username)
+            {
+                return array(
+                    'status'    => LOGIN_ERROR_USERNAME,
+                    'error_msg' => 'LOGIN_ERROR_USERNAME',
+                    'user_row'  => array('user_id' => ANONYMOUS),
+                );
+            }
+
+            $username_clean = utf8_clean_string($username);
+
+            $sql = 'SELECT user_id, username, user_password, user_passchg, user_pass_convert, user_email, user_type, user_login_attempts
+                FROM ' . USERS_TABLE . "
+                WHERE username_clean = '" . $this->db->sql_escape($username_clean) . "'";
+            $result = $this->db->sql_query($sql);
+            $row = $this->db->sql_fetchrow($result);
+            $this->db->sql_freeresult($result);
+
+            // Successful login... set user_login_attempts to zero...
+            return array(
+                'status'    => LOGIN_SUCCESS,
+                'error_msg' => false,
+                'user_row'  => $row,
+            );
+        }
+    }
+
+The service file
+++++++++++++++++
+For proper `dependency injection <https://wiki.phpbb.com/Dependency_Injection_Container>`_
```

Comment:
Would update once its removed from wiki (its a dev thing so that's up to the dev team to move eventually)